### PR TITLE
Fix: no redeclaration or re-use of types and functions

### LINKED ISSUE #15: Handle pre-declared opaque structure types
e.g. 
```ll
%ExecutableCache = type opaque
%ExecutableCache.2 = type opaque ; re-declared & used
%Array = type opaque
%Qubit = type opaque
%Result = type opaque
%String = type opaque
%Executable = type opaque
%ExecutionResult = type opaque
%Executable.1 = type opaque ; re-declared & used
%ExecutionResult.0 = type opaque ; re-declared & used
```

`qcs-sdk-qir` should probably look to see if any of these have been declared prior to its `transform` step. If known type identifier is encountered, it seems like it should be an error, but currently it is accepted and a new declaration is made.

### PR DESCRIPTION
**Pull request by nilslice**
Fixes #15.

There may be cases where QCS-related passes involving QIR transformations run more than once, and in these passes we currently re-build calls to functions or re-declare types. I think the worst case scenario here is that a function like `populate_executable_array` could be called multiple times and add real work. However, even though the possibly large build-up of repeat declarations is at minimum confusing, it could be a bit insidious.

### CODE REVIEW

File: src/context/types.rs [hunk 1]
```diff
@@ -30,22 +35,31 @@ pub(crate) struct Types<'ctx> {
...
+            Some(s) => s,
+            None => self.execution_result,
+        }
     }
 
     pub(crate) fn new(context: &'ctx Context) -> Self {
-        let execution_result = context.opaque_struct_type("ExecutionResult");
-        let executable = context.opaque_struct_type("Executable");
-        let executable_cache = context.opaque_struct_type("ExecutableCache");
+        let execution_result = context.opaque_struct_type(TYPE_NAME_EXECUTION_RESULT);
```
Comment (nilslice) [inline src/context/types.rs:60]: If/when this is merge-worthy, I'd like to take a second pass and re-structure some of the internal code organization so these const names for struct types and functions are more colocated and maybe behind some enum for each w/ conversion traits. e.g. 

```rust
enum QcsStructType {
    ExecutionResult,
    // ...
}

enum QcsFunction {
   PopulateExecutableArray,
   ExecuteOnQpu,
   // ...
}
```

File: src/transform/shot_count_block/qir.rs [hunk 2]
```diff
@@ -117,7 +126,14 @@ pub(crate) fn transpile_module(context: &mut QCSCompilerContext) -> Result<()> {
         .ok_or_else(|| eyre!("entrypoint function has no basic blocks"))?;
 
     match entry_basic_block.get_first_instruction() {
-        Some(instruction) => context.builder.position_before(&instruction),
+        Some(instruction) => {
+            if let Ok(Some(name)) = get_called_function_name(instruction) {
+                if name == "populate_executable_array" {
```
Comment (kalzoo) [inline src/transform/shot_count_block/qir.rs:131]: Why this check?

Also, want to use the const value in comparison?

File: src/transform/shot_count_block/qir.rs [hunk 3]
```diff
@@ -405,5 +421,6 @@ mod test {
         make_snapshot_test!(reduction);
         make_snapshot_test!(vqe_iteration);
         make_snapshot_test!(cartesian_rotations);
+        make_snapshot_test!(transformed_with_qcs_decls);
```
Comment (kalzoo) [inline src/transform/shot_count_block/qir.rs:424]:
```suggestion
        make_snapshot_test!(constant_parameter_values);
```

Looks like this was a holdover from an earlier change, but as you add it and its snapshots please update the name to something expressive

Comment (nilslice) [reply src/transform/shot_count_block/qir.rs]: 👌 neither are in tree after rebase -- good eye

File: src/transform/shot_count_block/qir.rs (hunk 2 above)
Comment (nilslice) [reply src/transform/shot_count_block/qir.rs]: good catch.. this looks like it was just an initial test to see if this part of the code had an effect on where any duplication may have occurred

**Review by kalzoo — APPROVED** [review]

**Pull request merged**